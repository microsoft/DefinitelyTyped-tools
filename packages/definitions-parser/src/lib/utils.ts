import * as https from "https";

export function getUrlContentsAsString(url: string): Promise<string> {
  return new Promise<string>((resolve, reject) => {
    https
      .get(url, (res) => {
        let data = "";
        res.on("data", (d) => (data += d));
        res.on("end", () => {
          resolve(data);
        });
      })
      .on("error", reject);
  });
}

export function withCache<T>(expiresInMs: number, getValue: () => Promise<T>): () => Promise<T> {
  let value: T | undefined;
  let resolvedAt: number | undefined;
  return async () => {
    if (resolvedAt === undefined || Date.now() - resolvedAt > expiresInMs) {
      value = await getValue();
      resolvedAt = Date.now();
    }
    return value!;
  };
}
