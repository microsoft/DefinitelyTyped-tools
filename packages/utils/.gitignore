logs
test/data/pack.tgz