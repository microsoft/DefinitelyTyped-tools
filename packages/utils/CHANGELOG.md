# @definitelytyped/utils

## 0.0.188

### Patch Changes

- Updated dependencies [8288affb]
  - @definitelytyped/typescript-versions@0.0.182

## 0.0.187

### Patch Changes

- 85379bf8: Fix writeJson helper, path and contents swapped

## 0.0.186

### Patch Changes

- 5b0559f6: Update and clean up dependencies
- Updated dependencies [5b0559f6]
  - @definitelytyped/typescript-versions@0.0.181

## 0.0.185

### Patch Changes

- 5d83a8ed: Fix loading on Windows
- 5d83a8ed: Fix command execution, paths on Windows

## 0.0.184

### Patch Changes

- a18ce6b1: Remove use of module resolution and OTHER_FILES, instead include all dts files in packages

  Files in packages are no longer determined by import resolution stemming from `index.d.ts` and tests (along with those listed in `OTHER_FILES.txt`).
  Instead, all files matching the glob `**/*.d.{ts,cts,mts,*.d.ts}` are included in the package, excluding those inside of versioned subdirectories.

  While not used for automated package publishing, an `.npmignore` is now required in each package.
  This allows for one-off `npm pack`-ing of packages, such that external tooling can get a rough approximation of what will be published for analysis.

## 0.0.183

### Patch Changes

- Updated dependencies [90e1d0ae]
  - @definitelytyped/typescript-versions@0.0.180

## 0.0.182

### Patch Changes

- 2b3138a0: Throw parse errors early now that more stuff is lazy

## 0.0.181

### Patch Changes

- 97f68d6e: Fix one more missing corepack issue, env in exec util

## 0.0.180

### Patch Changes

- d01cacd5: Make AllPackages lazy and asynchronous

## 0.0.179

### Patch Changes

- 024c3e73: Update @definitelytyped for Definitely Typed's monorepo upgrade
- 9fffa8ff: Add `compact` collection utility
- Updated dependencies [024c3e73]
  - @definitelytyped/typescript-versions@0.0.179
