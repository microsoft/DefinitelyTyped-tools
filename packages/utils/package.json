{
  "name": "@definitelytyped/utils",
  "version": "0.0.166-next.3",
  "description": "Shared utilities for DefinitelyTyped tools",
  "homepage": "https://github.com/microsoft/DefinitelyTyped-tools/tree/master/packages/utils#readme",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "repository": {
    "type": "git",
    "url": "https://github.com/microsoft/DefinitelyTyped-tools.git",
    "directory": "packages/utils"
  },
  "scripts": {
    "build": "tsc -b",
    "test": "../../node_modules/.bin/jest --config ../../jest.config.js packages/utils"
  },
  "bugs": {
    "url": "https://github.com/microsoft/DefinitelyTyped-tools/issues"
  },
  "dependencies": {
    "@definitelytyped/typescript-versions": "^0.0.166-next.3",
    "@qiwi/npm-registry-client": "^8.9.1",
    "@types/node": "^14.14.35",
    "charm": "^1.0.2",
    "fs-extra": "^8.1.0",
    "fstream": "^1.0.12",
    "tar": "^6.1.11",
    "tar-stream": "^2.1.4"
  },
  "devDependencies": {
    "@types/charm": "^1.0.1",
    "@types/fs-extra": "^8.1.0",
    "@types/tar-stream": "^2.1.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
