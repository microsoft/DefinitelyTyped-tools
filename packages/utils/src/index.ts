export * from "./assertions";
export * from "./async";
export * from "./collections";
export * from "./fs";
export * from "./io";
export * from "./logging";
export * from "./miscellany";
export * from "./npm";
export * from "./process";
export * from "./progress";
export * from "./typescript-installer";
