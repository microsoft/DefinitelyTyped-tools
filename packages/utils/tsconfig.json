{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "dist"
  },
  "include": ["src"],
  "references": [
    { "path": "../typescript-versions" },
    { "path": "../typescript-packages" }
  ]
}
