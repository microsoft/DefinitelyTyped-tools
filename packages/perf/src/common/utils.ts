import * as os from "os";
import * as fs from "fs";
import { randomBytes, createHash } from "crypto";
import { promisify } from "util";
import { execSync } from "child_process";
import { SystemInfo, Document, JSONDocument, PackageBenchmarkSummary, QueryResult } from "./types";
import { execAndThrowErrors } from "@definitelytyped/utils";
import {
  PackageId,
  gitChanges,
  formatDependencyVersion,
  parseVersionFromDirectoryName,
  TypingVersion
} from "@definitelytyped/definitions-parser";

export const pathExists = promisify(fs.exists);

export function ensureExists(...pathNames: string[]): string {
  for (const pathName of pathNames) {
    // tslint:disable-next-line:non-literal-fs-path -- filename comes from Azure artifact
    if (fs.existsSync(pathName)) {
      return pathName;
    }
  }
  const pathNamesPrint = pathNames.length > 1 ? "\n" + pathNames.map(s => ` - ${s}`).join("\n") : `'${pathNames[0]}`;
  throw new Error(`File or directory does not exist: ${pathNamesPrint}`);
}

export interface Args {
  [key: string]: string | boolean | number;
}

export function deserializeArgs(args: string[]): Args {
  const obj: Args = {};
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg.startsWith("--")) {
      const nextArg = args[i + 1];
      if (!nextArg || nextArg.startsWith("--")) {
        obj[arg.slice(2)] = true;
      } else {
        const numberArg = parseFloat(nextArg);
        const boolArg = nextArg === "true" ? true : nextArg === "false" ? false : undefined;
        obj[arg.slice(2)] = typeof boolArg === "boolean" ? boolArg : isNaN(numberArg) ? nextArg : numberArg;
        i++;
      }
    }
  }
  return obj;
}

export function serializeArgs(args: Args): string {
  return Object.keys(args)
    .map(arg => `--${arg}` + (args[arg] === true ? "" : args[arg].toString()))
    .join(" ");
}

export function compact<T>(arr: (T | null | undefined)[]): T[] {
  return arr.filter((elem): elem is T => elem !== null && elem !== undefined);
}

export function assertString(input: any, name?: string): string {
  if (typeof input !== "string") {
    throw new Error(`Expected a string for input${name ? ` '${name}'` : ""} but received ${typeof input}`);
  }
  return input;
}

export function assertNumber(input: any, name?: string): number {
  if (typeof input !== "number") {
    throw new Error(`Expected a number for input${name ? ` '${name}'` : ""} but received ${typeof input}`);
  }
  return input;
}

export function assertBoolean(input: any, name?: string): boolean {
  if (typeof input !== "boolean") {
    throw new Error(`Expected a boolean for input${name ? ` '${name}'` : ""} but received ${typeof input}`);
  }
  return input;
}

export function withDefault<T>(input: T, defaultValue: T): T {
  return input === undefined ? defaultValue : input;
}

export function getSystemInfo(): SystemInfo {
  const info = {
    cpus: os.cpus().map(({ times, ...cpu }) => cpu),
    arch: os.arch(),
    platform: os.platform(),
    release: os.release(),
    totalmem: os.totalmem(),
    nodeVersion: process.version
  };

  return {
    ...info,
    hash: createHash("md5")
      .update(JSON.stringify(info))
      .digest("hex")
  };
}

export interface GetChangedPackagesOptions {
  diffFrom?: string;
  diffTo: string;
  definitelyTypedPath: string;
}

export async function getChangedPackages({
  diffFrom = "HEAD",
  diffTo,
  definitelyTypedPath
}: GetChangedPackagesOptions) {
  const diff = await execAndThrowErrors(`git diff --name-status ${diffFrom} ${diffTo}`, definitelyTypedPath);
  if (!diff) {
    return undefined;
  }

  const changes = diff.split("\n").map(line => {
    const [status, file] = line.split(/\s+/, 2);
    return { status: status.trim() as "A" | "D" | "M", file: file.trim() };
  });

  return gitChanges(changes);
}

export function packageIdsAreEqual(a: PackageId): (b: PackageId) => boolean;
export function packageIdsAreEqual(a: PackageId, b: PackageId): boolean;
export function packageIdsAreEqual(a: PackageId, b?: PackageId): boolean | ((b: PackageId) => boolean) {
  return b === undefined ? equalsA : equalsA(b);
  function equalsA(b: PackageId) {
    return a.name === b.name && a.version === b.version;
  }
}

export function getPercentDiff(actual: number, expected: number): number {
  return (actual - expected) / expected;
}

export function isWithin(actual: number, expected: number, tolerance: number): boolean {
  return Math.abs(getPercentDiff(actual, expected)) <= tolerance;
}

export function systemsAreCloseEnough(a: SystemInfo, b: SystemInfo, cpuSpeedTolerance = 0.1): boolean {
  if (a.hash === b.hash) {
    return true;
  }
  return (
    a.arch === b.arch &&
    a.platform === b.platform &&
    a.nodeVersion === b.nodeVersion &&
    a.cpus.length === b.cpus.length &&
    a.cpus.every((cpu, index) => {
      const otherCPU = b.cpus[index];
      return cpu.model === otherCPU.model && isWithin(cpu.speed, otherCPU.speed, cpuSpeedTolerance);
    })
  );
}

export function createDocument<T>(body: T, version: number): Document<T> {
  return {
    version,
    createdAt: new Date(),
    system: getSystemInfo(),
    body
  };
}

export function parsePackageKey(key: string): PackageId {
  const [name, versionString] = key.split("/");
  const version = parseVersionFromDirectoryName(versionString);
  return {
    name,
    version: version || ("*" as const)
  };
}

export function toPackageKey(name: string, version: string | TypingVersion): string;
export function toPackageKey(packageId: PackageId): string;
export function toPackageKey(packageIdOrName: string | PackageId, version?: string | TypingVersion) {
  const packageId =
    typeof packageIdOrName === "string"
      ? {
          name: packageIdOrName,
          version: (typeof version === "string" ? parseVersionFromDirectoryName(version) : version) || ("*" as const)
        }
      : packageIdOrName;
  return `${packageId.name}/${formatDependencyVersion(packageId.version)}`;
}

export function deserializeSummary(
  doc: QueryResult<JSONDocument<PackageBenchmarkSummary>>
): QueryResult<Document<PackageBenchmarkSummary>>;
export function deserializeSummary(doc: JSONDocument<PackageBenchmarkSummary>): Document<PackageBenchmarkSummary> {
  return {
    ...doc,
    createdAt: new Date(doc.createdAt),
    body: {
      ...doc.body,
      batchRunStart: new Date(doc.body.batchRunStart)
    }
  };
}

export function getSourceVersion(cwd: string) {
  return execSync("git rev-parse HEAD", { cwd, encoding: "utf8" }).trim();
}

export function shuffle<T>(array: readonly T[]): T[] {
  const output = array.slice();
  let counter = output.length;
  while (counter > 0) {
    const index = Math.floor((randomBytes(1).readUInt8(0) / 2 ** 8) * counter);
    counter--;
    const elem = output[counter];
    output[counter] = output[index];
    output[index] = elem;
  }

  return output;
}

export function not<P extends unknown[]>(fn: (...args: P) => boolean): (...args: P) => boolean {
  return (...args) => !fn(...args);
}

export function findLast<T>(arr: T[], predicate: (element: T) => boolean): T | undefined {
  for (let i = arr.length - 1; i >= 0; i--) {
    const element = arr[i];
    if (predicate(element)) {
      return element;
    }
  }
  return;
}
