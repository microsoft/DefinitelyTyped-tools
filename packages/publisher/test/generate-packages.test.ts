import {
  createNotNeededPackageJSON,
  createPackageJSON,
  createReadme,
  getLicenseFileText
} from "../src/generate-packages";
import {
  AllPackages,
  License,
  NotNeededPackage,
  readNotNeededPackages,
  TypesDataFile,
  TypingsData,
  TypingsDataRaw,
  createMockDT
} from "@definitelytyped/definitions-parser";
import { testo } from "./utils";
import { Registry } from "@definitelytyped/utils";

function createRawPackage(license: License): TypingsDataRaw {
  return {
    libraryName: "jquery",
    typingsPackageName: "jquery",
    dependencies: { madeira: { major: 1 } },
    testDependencies: [],
    pathMappings: [],
    contributors: [{ name: "A", url: "b@c.d", githubUsername: "e" }],
    libraryMajorVersion: 1,
    libraryMinorVersion: 0,
    minTsVersion: "3.2",
    typesVersions: [],
    files: ["index.d.ts", "jquery.test.ts"],
    license,
    packageJsonDependencies: [{ name: "balzac", version: "~3" }],
    contentHash: "11",
    projectName: "jquery.org",
    globals: [],
    declaredModules: ["jquery"]
  };
}
function createTypesData(): TypesDataFile {
  return {
    jquery: {
      1: createRawPackage(License.MIT)
    },
    madeira: {
      1: createRawPackage(License.Apache20)
    }
  };
}
function createUnneededPackage() {
  return new NotNeededPackage({
    libraryName: "absalom",
    typingsPackageName: "absalom",
    asOfVersion: "1.1.1",
    sourceRepoURL: "https://github.com/aardwulf/absalom"
  });
}
testo({
  mitLicenseText() {
    const typing = new TypingsData(createRawPackage(License.MIT), /*isLatest*/ true);
    expect(getLicenseFileText(typing)).toEqual(expect.stringContaining("MIT License"));
  },
  apacheLicenseText() {
    const typing = new TypingsData(createRawPackage(License.Apache20), /*isLatest*/ true);
    expect(getLicenseFileText(typing)).toEqual(expect.stringContaining("Apache License, Version 2.0"));
  },
  basicReadme() {
    const typing = new TypingsData(createRawPackage(License.Apache20), /*isLatest*/ true);
    expect(createReadme(typing)).toEqual(expect.stringContaining("This package contains type definitions for"));
  },
  readmeContainsProjectName() {
    const typing = new TypingsData(createRawPackage(License.Apache20), /*isLatest*/ true);
    expect(createReadme(typing)).toEqual(expect.stringContaining("jquery.org"));
  },
  readmeOneDependency() {
    const typing = new TypingsData(createRawPackage(License.Apache20), /*isLatest*/ true);
    expect(createReadme(typing)).toEqual(
      expect.stringContaining("Dependencies: [@types/madeira](https://npmjs.com/package/@types/madeira)")
    );
  },
  readmeNoGlobals() {
    const typing = new TypingsData(createRawPackage(License.Apache20), /*isLatest*/ true);
    expect(createReadme(typing)).toEqual(expect.stringContaining("Global values: none"));
  },
  basicPackageJson() {
    const packages = AllPackages.from(createTypesData(), readNotNeededPackages(createMockDT().fs));
    const typing = new TypingsData(createRawPackage(License.MIT), /*isLatest*/ true);
    expect(createPackageJSON(typing, "1.0", packages, Registry.NPM)).toEqual(`{
    "name": "@types/jquery",
    "version": "1.0",
    "description": "TypeScript definitions for jquery",
    "license": "MIT",
    "contributors": [
        {
            "name": "A",
            "url": "b@c.d",
            "githubUsername": "e"
        }
    ],
    "main": "",
    "types": "index.d.ts",
    "repository": {
        "type": "git",
        "url": "https://github.com/DefinitelyTyped/DefinitelyTyped.git",
        "directory": "types/jquery"
    },
    "scripts": {},
    "dependencies": {
        "@types/madeira": "^1",
        "balzac": "~3"
    },
    "typesPublisherContentHash": "11",
    "typeScriptVersion": "3.2"
}`);
  },
  githubPackageJsonName() {
    const packages = AllPackages.from(createTypesData(), readNotNeededPackages(createMockDT().fs));
    const typing = new TypingsData(createRawPackage(License.MIT), /*isLatest*/ true);
    expect(createPackageJSON(typing, "1.0", packages, Registry.Github)).toEqual(
      expect.stringContaining('"name": "@types/jquery"')
    );
  },
  githubPackageJsonRegistry() {
    const packages = AllPackages.from(createTypesData(), readNotNeededPackages(createMockDT().fs));
    const typing = new TypingsData(createRawPackage(License.MIT), /*isLatest*/ true);
    const s = createPackageJSON(typing, "1.0", packages, Registry.Github);
    expect(s).toEqual(expect.stringContaining("publishConfig"));
    expect(s).toEqual(expect.stringContaining('"registry": "https://npm.pkg.github.com/"'));
  },
  basicNotNeededPackageJson() {
    const s = createNotNeededPackageJSON(createUnneededPackage(), Registry.NPM);
    expect(s).toEqual(`{
    "name": "@types/absalom",
    "version": "1.1.1",
    "typings": null,
    "description": "Stub TypeScript definitions entry for absalom, which provides its own types definitions",
    "main": "",
    "scripts": {},
    "author": "",
    "repository": "https://github.com/aardwulf/absalom",
    "license": "MIT",
    "dependencies": {
        "absalom": "*"
    }
}`);
  },
  scopedNotNeededPackageJson() {
    const scopedUnneeded = new NotNeededPackage({
      libraryName: "@google-cloud/pubsub",
      typingsPackageName: "google-cloud__pubsub",
      asOfVersion: "0.26.0",
      sourceRepoURL: "https://github.com/googleapis/nodejs-storage"
    });
    const s = createNotNeededPackageJSON(scopedUnneeded, Registry.NPM);
    expect(s).toEqual(`{
    "name": "@types/google-cloud__pubsub",
    "version": "0.26.0",
    "typings": null,
    "description": "Stub TypeScript definitions entry for @google-cloud/pubsub, which provides its own types definitions",
    "main": "",
    "scripts": {},
    "author": "",
    "repository": "https://github.com/googleapis/nodejs-storage",
    "license": "MIT",
    "dependencies": {
        "@google-cloud/pubsub": "*"
    }
}`);
  },
  githubNotNeededPackageJson() {
    const s = createNotNeededPackageJSON(createUnneededPackage(), Registry.Github);
    expect(s).toEqual(expect.stringContaining("@types"));
    expect(s).toEqual(expect.stringContaining("npm.pkg.github.com"));
  }
});
