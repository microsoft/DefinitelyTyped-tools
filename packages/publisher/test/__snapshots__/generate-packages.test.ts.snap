// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apacheLicenseText 1`] = `
"Copyright 2024 A, E
Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License."
`;

exports[`basicNotNeededPackageJson 1`] = `
"{
    "name": "@types/absalom",
    "version": "1.1.1",
    "description": "Stub TypeScript definitions entry for alternate, which provides its own types definitions",
    "main": "",
    "scripts": {},
    "license": "MIT",
    "dependencies": {
        "alternate": "*"
    },
    "deprecated": "This is a stub types definition. alternate provides its own type definitions, so you do not need this installed."
}"
`;

exports[`basicPackageJson 1`] = `
"{
    "name": "@types/jquery",
    "version": "1.0",
    "description": "TypeScript definitions for jquery",
    "homepage": "https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jquery",
    "license": "MIT",
    "contributors": [
        {
            "name": "A",
            "url": "b@c.d"
        },
        {
            "name": "E",
            "githubUsername": "e",
            "url": "https://github.com/e"
        }
    ],
    "main": "",
    "types": "index.d.ts",
    "repository": {
        "type": "git",
        "url": "https://github.com/DefinitelyTyped/DefinitelyTyped.git",
        "directory": "types/jquery"
    },
    "scripts": {},
    "dependencies": {
        "@types/madeira": "^1"
    },
    "peerDependencies": {
        "@types/express": "*"
    },
    "typesPublisherContentHash": "53300522250468c4161b10d962cac2d9d8f2cfee1b3dfef4b749a7c3ec839275",
    "typeScriptVersion": "5.0"
}"
`;

exports[`mitLicenseText 1`] = `
"    MIT License

    Copyright (c) Microsoft Corporation.

    Permission is hereby granted, free of charge, to any person obtaining a copy
    of this software and associated documentation files (the "Software"), to deal
    in the Software without restriction, including without limitation the rights
    to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
    copies of the Software, and to permit persons to whom the Software is
    furnished to do so, subject to the following conditions:

    The above copyright notice and this permission notice shall be included in all
    copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
    AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
    OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
    SOFTWARE
"
`;

exports[`readmeContainsManyDTSFilesDoesNotAmendREADME 1`] = `
"# Installation
> \`npm install --save @types/jquery\`

# Summary
This package contains type definitions for jquery (jquery.org).

# Details
Files were exported from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jquery.

### Additional Details
 * Last updated: Mon, 09 Dec 2024 20:10:05 GMT
 * Dependencies: [@types/madeira](https://npmjs.com/package/@types/madeira)
 * Peer dependencies: [@types/express](https://npmjs.com/package/@types/express)

# Credits
These definitions were written by [A](b@c.d), and [E](https://github.com/e).
"
`;

exports[`readmeJquery 1`] = `
"# Installation
> \`npm install --save @types/jquery\`

# Summary
This package contains type definitions for jquery (jquery.org).

# Details
Files were exported from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jquery.
## [index.d.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jquery/index.d.ts)
\`\`\`\`ts
type T = import("./types");

\`\`\`\`

### Additional Details
 * Last updated: Mon, 09 Dec 2024 20:10:05 GMT
 * Dependencies: [@types/madeira](https://npmjs.com/package/@types/madeira)
 * Peer dependencies: [@types/express](https://npmjs.com/package/@types/express)

# Credits
These definitions were written by [A](b@c.d), and [E](https://github.com/e).
"
`;

exports[`readmeMultipleDependencies 1`] = `
"# Installation
> \`npm install --save @types/jquery\`

# Summary
This package contains type definitions for jquery (jquery.org).

# Details
Files were exported from https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jquery.
## [index.d.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/jquery/index.d.ts)
\`\`\`\`ts
type T = import("./types");

\`\`\`\`

### Additional Details
 * Last updated: Mon, 09 Dec 2024 20:10:05 GMT
 * Dependencies: [@types/example](https://npmjs.com/package/@types/example), [@types/madeira](https://npmjs.com/package/@types/madeira)
 * Peer dependencies: [@types/example2](https://npmjs.com/package/@types/example2), [@types/express](https://npmjs.com/package/@types/express)

# Credits
These definitions were written by [A](b@c.d), and [E](https://github.com/e).
"
`;

exports[`scopedNotNeededPackageJson 1`] = `
"{
    "name": "@types/google-cloud__pubsub",
    "version": "0.26.0",
    "description": "Stub TypeScript definitions entry for @google-cloud/chubdub, which provides its own types definitions",
    "main": "",
    "scripts": {},
    "license": "MIT",
    "dependencies": {
        "@google-cloud/chubdub": "*"
    },
    "deprecated": "This is a stub types definition. @google-cloud/chubdub provides its own type definitions, so you do not need this installed."
}"
`;
